README.md
# 獅尾B2宋朝 Swei B2 Serif

獅尾B2宋朝字體基於[思源宋體](https://github.com/adobe-fonts/source-han-serif)的三角形轉成半圓形改造，彷彿墨暈或是稍微過曝的效果。支援简体中文、繁體中文、韓文與日文；可以免費商用，歡迎大家自由應用、自由優化、自由改作！

和思源宋體一樣，支援7種的字重：
![字重比較](https://github.com/max32002/swei-b2-serif/raw/master/preview/compare_style.png)

## 與其他字體的比較
「源雲明體JP」(1)、「源雲明體TW」(2)、「思源宋體」(3)、「獅尾B2宋朝」(4)的比較：
![字體比較](https://github.com/max32002/swei-b2-serif/raw/master/preview/compare_noto_serif.png)

* 在「肉」、「糸」、「女」、「辶」、「食」的筆畫，獅尾比較相似台灣教育部推薦字體筆順，適合教育用途。
* 在「草」部的筆畫，獅尾是分開的二個部份。
* 字體裡的「草字頭部首」中間該不該連起來，「肉字旁部首」該不該變成「月」，要不要把手寫習慣代入印刷字體？整體看來，會不會影響視覺的延伸性？會不會影響印刷的可行性與閱讀的便利性？這個答案我也不清楚，如果你是台灣教育部標準字符的愛好者，獅尾字體應該是一個不錯的選擇。

### 字體後面的 SC,JP,TC是什意思？
* SC是 Simplified Chinese 简体中文，代表大陸習慣字形。
* TC是 Traditional Chinese 繁体中文，代表港台習慣的字形。
* JP是 Japanese 日文，代表日本習慣字形。
* 相同一個字，在不地區的書寫方式可能會略有不同。

### 「CJK TC」和「CJK SC」的比較：
![TC和SC比較](https://github.com/max32002/swei-b2-serif/raw/master/preview/compare_tc_sc.png)
* 在「肉」、「糸」、「女」、「辶」、「食」的筆畫不同，CJK SC 「肉字旁部首」變成「月」。 
* 在「草」部的筆畫，CJK SC 中間連起來。
* 「體」字的骨上方方向相反。
* 「角」字下面穿頭。
* 雨、身、戶、舟、北、㕣、酋字寫法不同。

### 「CJK TC」和「CJK JP」的比較：
![TC和SC比較](https://github.com/max32002/swei-b2-serif/raw/master/preview/compare_tc_jp.png)
* 在「肉」、「糸」、「女」、「辶」、「食」的筆畫不同，CJK JP 「肉字旁部首」變成「月」。 
* 在「草」部的筆畫，CJK JP 中間連起來。
* 雨、言、青、兌、賣、直、真、曾、戶、北、㕣、酋字寫法不同。其他[常見與日系字型的差異](https://max-everyday.com/2020/06/taiwanpearl/#diff)。

## 更新日誌
[點擊此處](https://github.com/max32002/swei-b2-serif/blob/master/change_log.md) 查看更新記錄。

## 下載字型

請點選GitHub此畫面右上綠色「Clone or download」按鈕，並選擇「Download ZIP」，或點進想下載的ttf字型檔案，再點「Download」的按鈕進行下載。

## 網頁字型(Web Font)服務

網頁字型用於網頁上的字型顯示，訪客不需預先安裝字型檔，一樣能夠看到特殊的字型效果。不只是電腦，在智慧型手機和平板裝置的瀏覽器上也可正常顯示。實現該功能的原理是在瀏覽時才下載字型檔。

可以服用下面的css:
```
@font-face {
  font-family: SweiB2SerifCJKtc-Regular;
  src: url(https://cdn.jsdelivr.net/gh/max32002/swei-b2-serif@1.0/WebFont/CJK%20TC/SweiB2SerifCJKtc-Regular.woff2) format("woff2")
  , url(https://cdn.jsdelivr.net/gh/max32002/swei-b2-serif@1.0/WebFont/CJK%20TC/SweiB2SerifCJKtc-Regular.woff) format("woff");
}
```

## 附註

* 這不是一個完整、健康和專業的字體。
* 由於小編對字體編碼方式完全不清楚，相較於原版的思源宋體，可能有掉很多符號或不常用的字。
* 演算黑科技將字體三角形換成圓形，請參考 /python/ 目錄下的腳本檔案。透過調整程式碼，也許也可以產生出新的有趣字型。目前的程式應該還有很多錯誤或需要再加強的地方。請先把要處理的字型，透過 FontForge 開啟，並另存 FontForge 的專案為資料夾格式(.sfdir)，最後就可以透過Max的 Python 程式去處理產生出來的檔案。

## 著作權與授權

* 本字型是基於 SIL Open Font License 1.1 改造Adobe和Google所開發、發表的「[思源宋體](https://github.com/adobe-fonts/source-han-serif)」字型。
* 本字型亦基於 SIL Open Font License 1.1 授權條款免費公開，關於授權合約的內容、免責事項等細節，請詳讀 License 文件。
    * 可自由商用 不需付費、知會或標明作者，即可自由使用此字型，亦可做商業應用。
    * 可自由傳布 可自由分享檔案、將檔案安裝於任何軟硬體中。
    * 可自由改作為其他字型 將字型檔案修改重製為其他字型檔案，改作後的字型檔案須同樣依 Open Font License 釋出。
    
    
## 相關網頁

花園家族：
* B2花園 B2 Hana
https://max-everyday.com/2020/08/b2-hana-font/
* 花園肉丸 Hana Meatball
https://max-everyday.com/2020/08/hana-meatball/

獅尾黑體家族：
* 獅尾鋸腿黑體 Swei Alias Leg
https://max-everyday.com/2020/10/swei-alias-leg/
* 獅尾鋸齒黑體 Swei Alias Sans
https://max-everyday.com/2020/10/swei-alias-sans/
* 獅尾尖腿黑體 Swei Spike Leg
https://max-everyday.com/2020/10/swei-spike-leg/
* 獅尾尖刺黑體 Swei Spike Sans
https://max-everyday.com/2020/10/swei-spike-sans/
* 獅尾快腿黑體 Swei Shear Leg
https://max-everyday.com/2020/09/swei-shear-leg/
* 獅尾快剪黑體 Swei Shear Sans
https://max-everyday.com/2020/09/swei-shear-sans/
* 獅尾福腿黑體 Swei Gospel Leg
https://max-everyday.com/2020/09/swei-gospel-leg/
* 獅尾福音黑體 Swei Gospel Sans
https://max-everyday.com/2020/09/swei-gospel-sans/
* 獅尾D滷腿黑體 Swei Del Luna Leg
https://max-everyday.com/2020/09/swei-del-luna-leg/
* 獅尾德魯納黑體 Swei Del Luna Sans
https://max-everyday.com/2020/09/swei-del-luna-sans/
* 獅尾彎腿黑體 Swei Curve Leg
https://max-everyday.com/2020/09/swei-curve-leg/
* 獅尾彎黑體 Swei Curve Sans
https://max-everyday.com/2020/09/swei-curve-sans/
* 獅尾霓腿黑體 Swei Bow Leg
https://max-everyday.com/2020/09/swei-bow-leg/
* 獅尾霓黑體 Swei Bow Sans
https://max-everyday.com/2020/09/swei-bow-sans/
* 獅尾蝙蝠圓體 Swei Bat Sans
https://max-everyday.com/2020/09/swei-bat-sans/
* 獅尾牙膏圓體 Swei Toothpaste
https://max-everyday.com/2020/09/swei-toothpaste/
* 獅尾三腿黑體 Swei 3T Leg
https://max-everyday.com/2020/09/swei-3t-leg/
* 獅尾三角黑體 Swei 3T Sans
https://max-everyday.com/2020/08/swei-3t-sans/
* 獅尾螺帽腿黑體 Swei Nut Leg
https://max-everyday.com/2020/08/swei-nut-leg/
* 獅尾螺帽黑體 Swei Nut Sans
https://max-everyday.com/2020/08/swei-nut-sans/
* 獅尾B2腿黑體 Swei B2 Leg
https://max-everyday.com/2020/07/swei-b2-leg/
* 獅尾B2黑體 Swei B2 Sans
https://max-everyday.com/2020/07/swei-b2-sans/
* 獅尾腿圓 Swei Gothic Leg
https://max-everyday.com/2020/08/swei-gothic-leg/
* 獅尾彩虹腿 Swei Rainbow Leg
https://max-everyday.com/2020/08/swei-rainbow-leg/
* 獅尾XD珍珠 Swei XD Pearl
https://max-everyday.com/2020/07/swei-xd-pearl/
* 獅尾D露西 Swei D Lucy
https://max-everyday.com/2020/07/swei-d-lucy/
* 獅尾半月字體 Swei Gothic
https://max-everyday.com/2020/04/swei-half-moon/
* 台灣圓體 TaiwanPearl
https://max-everyday.com/2020/06/taiwanpearl/
* 獅尾圓體 Swei Gothic
https://max-everyday.com/2020/04/swei-gothic/
* 獅尾黑體 Swei Sans
https://max-everyday.com/2020/03/swei-sans/

獅尾宋體家族：
* 獅尾B2宋朝 Swei B2 Serif
https://max-everyday.com/2020/07/swei-b2-serif/
* 獅尾肉丸 Swei Meatball
https://max-everyday.com/2020/06/swei-meatball/
* 獅尾四季春字體 Swei Spring
https://max-everyday.com/2020/04/swei-spring/

其他字體：
* 何某手寫體 Nani Font
https://max-everyday.com/2020/09/nanifont/
* 內海字體  Naikai Font
https://max-everyday.com/2020/03/naikaifont/
* 莫大毛筆字體 Bakudai Font
https://max-everyday.com/2020/03/bakudaifont/
* 正風毛筆字體 Masa Font
https://max-everyday.com/2020/05/masafont/
* 假粉圓體 Fake Pearl 
https://max-everyday.com/2020/03/open-huninn-font/
* 俊羽圓體 Yu Pearl 
https://max-everyday.com/2020/03/yupearl/

其他網站：
* 清松手寫體 JasonHandWriting
https://jasonfonts.max-everyday.com/
* Max學習字體相關的筆記
https://codereview.max-everyday.com/font-readme/

## 贊助Max

很高興可以替中華民國美學盡一分心力、讓台灣擁有更好的文字風景，希望能提供另一種美學讓大家選擇，如果你覺得這篇文章寫的很好，想打賞Max，贊助方式如下：
https://max-everyday.com/about/#donate


change_log.md
# 2020/10/12 ver 1.080
* 支援直排、直書。

# 2020/07/23 ver 1.078
* 增加套用到的效果的點。

# 2020/07/16 ver 1.076
* 增加 CJK SC 中國大陸地區版本。

# 2020/07/16 ver 1.070
* 第1版。

python/util/Rule11_Inside_Curve.py
#!/usr/bin/env python3
#encoding=utf-8

from . import spline_util
from . import Rule
import math

# RULE # 11
# check inside curve
# PS: 因為 array size change, so need redo.
class Rule(Rule.Rule):
    def __init__(self):
        pass

    def apply(self, spline_dict, resume_idx, inside_stroke_dict, skip_coordinate, skip_coordinate_rule):
        redo_travel=False
        check_first_point = False

        # 最大的角度值，超過就skip
        ALMOST_LINE_RATE = 1.84

        # default: 1.75
        SLIDE_1_PERCENT_MIN = 0.8
        SLIDE_1_PERCENT_MAX = ALMOST_LINE_RATE

        MIN_DISTANCE = 12

        # clone
        format_dict_array=[]
        format_dict_array = spline_dict['dots'][1:]
        format_dict_array = self.caculate_distance(format_dict_array)

        nodes_length = len(format_dict_array)
        #print("orig nodes_length:", len(spline_dict['dots']))
        #print("format nodes_length:", nodes_length)
        #print("resume_idx:", resume_idx)

        rule_need_lines = 3
        fail_code = -1
        if nodes_length >= rule_need_lines:
            for idx in range(nodes_length):
                if idx <= resume_idx:
                    # skip traveled nodes.
                    continue

                is_debug_mode = False
                #is_debug_mode = True

                if [format_dict_array[idx]['x'],format_dict_array[idx]['y']] in skip_coordinate:
                    if is_debug_mode:
                        print("match skip idx+0:",format_dict_array[(idx+0)%nodes_length]['code'])
                        pass
                    continue

                # 要轉換的原來的角，第4點，不能就是我們產生出來的曲線結束點。
                # for case.3122 上面的點。
                if [format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y']] in skip_coordinate:
                    if is_debug_mode:
                        print("match skip idx+2:",format_dict_array[(idx+2)%nodes_length]['code'])
                        pass
                    continue

                if format_dict_array[idx]['code'] in skip_coordinate_rule:
                    if is_debug_mode:
                        print("match skip skip_coordinate_rule +0:",format_dict_array[idx]['code'])
                        pass
                    continue

                is_debug_mode = False
                #is_debug_mode = True

                if is_debug_mode:
                    debug_coordinate_list = [[463,602]]
                    #print("="*30)
                    #print("current x,y:", [format_dict_array[idx]['x'],format_dict_array[idx]['y']])
                    if not([format_dict_array[idx]['x'],format_dict_array[idx]['y']] in debug_coordinate_list):
                        continue
                        pass
                    else:
                        print("="*30)
                        print("index:", idx)
                        for debug_idx in range(8):
                            print(debug_idx-2,": #11 val:",format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['code'],'-(',format_dict_array[(idx+debug_idx+nodes_length-2)%nodes_length]['distance'],')')

                is_match_pattern = False

                x0 = format_dict_array[(idx+0)%nodes_length]['x']
                y0 = format_dict_array[(idx+0)%nodes_length]['y']
                x1 = format_dict_array[(idx+1)%nodes_length]['x']
                y1 = format_dict_array[(idx+1)%nodes_length]['y']
                x2 = format_dict_array[(idx+2)%nodes_length]['x']
                y2 = format_dict_array[(idx+2)%nodes_length]['y']

                # use more close coordinate.
                if format_dict_array[(idx+0)%nodes_length]['code']=='c':
                    x0 = format_dict_array[(idx+0)%nodes_length]['x2']
                    y0 = format_dict_array[(idx+0)%nodes_length]['y2']
                if format_dict_array[(idx+2)%nodes_length]['code']=='c':
                    x2 = format_dict_array[(idx+2)%nodes_length]['x1']
                    y2 = format_dict_array[(idx+2)%nodes_length]['y1']

                previous_x,previous_y=0,0
                next_x,next_y=0,0

                # match ?l?
                if format_dict_array[(idx+1)%nodes_length]['t'] == 'l':
                    fail_code = 100
                    is_match_pattern = True

                # match ??l
                if format_dict_array[(idx+2)%nodes_length]['t'] == 'l':
                    fail_code = 110
                    is_match_pattern = True

                # compare distance, muse large than our "large round"
                if is_match_pattern:
                    fail_code = 200
                    is_match_pattern = False
                    if format_dict_array[(idx+0)%nodes_length]['distance'] >= MIN_DISTANCE:
                        if format_dict_array[(idx+1)%nodes_length]['distance'] >= MIN_DISTANCE:
                            is_match_pattern = True

                # for D.Lucy
                if self.config.PROCESS_MODE in ["D"]:
                    if is_match_pattern:
                        is_match_d_base_rule, fail_code = self.going_d_right(format_dict_array,idx)
                        is_match_pattern = is_match_d_base_rule

                previous_x,previous_y=0,0
                next_x,next_y=0,0

                # skip small angle
                if is_match_pattern:
                    fail_code = 300
                    is_match_pattern = False

                    # PS: 使用結束 x,y，會造成更誤判，因為沒有另外儲存 rule#99 處理記錄，會造成重覆套用。
                    #slide_percent_1 = spline_util.slide_percent(format_dict_array[(idx+0)%nodes_length]['x'],format_dict_array[(idx+0)%nodes_length]['y'],format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'])
                    slide_percent_1 = spline_util.slide_percent(x0,y0,x1,y1,x2,y2)

                    if is_debug_mode:
                            print("slide_percent 1:", slide_percent_1)
                            print("data end:",format_dict_array[(idx+0)%nodes_length]['x'],format_dict_array[(idx+0)%nodes_length]['y'],format_dict_array[(idx+1)%nodes_length]['x'],format_dict_array[(idx+1)%nodes_length]['y'],format_dict_array[(idx+2)%nodes_length]['x'],format_dict_array[(idx+2)%nodes_length]['y'])
                            print("data virtual:",x0,y0,x1,y1,x2,y2)
                            print("SLIDE_1_PERCENT_MIN:", SLIDE_1_PERCENT_MIN)
                            print("SLIDE_1_PERCENT_MAX:", SLIDE_1_PERCENT_MAX)
        
                    if slide_percent_1 >= SLIDE_1_PERCENT_MIN and slide_percent_1 <= SLIDE_1_PERCENT_MAX:
                        is_match_pattern = True
                    else:
                        # try real point.
                        # for case "加"的力的右上角。

                        # try real point.
                        # for case 「加」字的力的右上角。
                        # PS: 「加」字算是例外，一般的字，不應檢查到這裡。
                        if format_dict_array[(idx+1)%nodes_length]['t']=='c':
                            if format_dict_array[(idx+1)%nodes_length]['x2']==format_dict_array[(idx+1)%nodes_length]['x1'] and format_dict_array[(idx+1)%nodes_length]['y2']==format_dict_array[(idx+1)%nodes_length]['y1']:
                                pass
                            else:
                                # 這個情況，滿特別的，就允許例外看看。
                                x0 = format_dict_array[(idx+0)%nodes_length]['x']
                                y0 = format_dict_array[(idx+0)%nodes_length]['y']
                        x1 = format_dict_array[(idx+1)%nodes_length]['x']
                        y1 = format_dict_array[(idx+1)%nodes_length]['y']
                        
                        if format_dict_array[(idx+2)%nodes_length]['t']=='c':
                            if format_dict_array[(idx+2)%nodes_length]['x2']==format_dict_array[(idx+2)%nodes_length]['x1'] and format_dict_array[(idx+2)%nodes_length]['y2']==format_dict_array[(idx+2)%nodes_length]['y1']:
                                pass
                            else:
                                # 這個情況，滿特別的，就允許例外看看。
                                x2 = format_dict_array[(idx+2)%nodes_length]['x']
                                y2 = format_dict_array[(idx+2)%nodes_length]['y']
                        
                        slide_percent_1 = spline_util.slide_percent(x0,y0,x1,y1,x2,y2)
                        if slide_percent_1 >= SLIDE_1_PERCENT_MIN and slide_percent_1 <= SLIDE_1_PERCENT_MAX:
                            is_match_pattern = True

                # check black stroke in white area. @_@;
                is_apply_large_corner = False
                if is_match_pattern:
                    inside_stroke_flag,inside_stroke_dict = self.test_inside_coner(x0, y0, x1, y1, x2, y2, self.config.STROKE_MIN, inside_stroke_dict)
                    if inside_stroke_flag:
                        is_apply_large_corner = True
                        #print("match is_apply_large_corner:",x1,y1)

                if not is_apply_large_corner:
                    if is_match_pattern:
                        fail_code = 400
                        is_match_pattern = False
                        
                        join_line_debug_mode = False      # online
                        #join_line_debug_mode = True       # debug
                        join_flag = self.join_line_check(x0,y0,x1,y1,x2,y2,debug_mode=join_line_debug_mode)
                        join_flag_1 = join_flag
                        join_flag_2 = None
                        if not join_flag:
                            join_flag = self.join_line_check(x2,y2,x1,y1,x0,y0)
                            join_flag_2 = join_flag

                        if is_debug_mode:
                            print("check1_flag:",join_flag_1, "data:", x0,y0,x1,y1,x2,y2)
                            print("check2_flag:",join_flag_2, "data:", x2,y2,x1,y1,x0,y0)
                            print("final join flag:", join_flag)
                        
                        if not join_flag:
                            #print("match small coner")
                            #print(idx,"small rule5:",format_dict_array[idx]['code'])
                            is_match_pattern = True
                            #is_apply_small_corner = True
                            #pass


                if is_debug_mode:
                    if not is_match_pattern:
                        print(idx,"debug fail_code #11:", fail_code)
                    else:
                        print("match rule #11:",idx)

                if is_match_pattern:
                    #print("match rule #11")
                    #print(idx,"debug rule11+0:",format_dict_array[idx]['code'])
                    #print(idx,"debug rule11+1:",format_dict_array[(idx+1)%nodes_length]['code'])
                    #print(idx,"debug rule11+2:",format_dict_array[(idx+2)%nodes_length]['code'])

                    # to avoid same code apply twice.
                    nodes_length = len(format_dict_array)
                    generated_code = format_dict_array[(idx+0)%nodes_length]['code']
                    #print("generated_code:", generated_code)
                    skip_coordinate_rule.append(generated_code)

                    # make coner curve
                    round_offset = self.config.OUTSIDE_ROUND_OFFSET
                    # large curve, use small angle.
                    # start to resize round offset size.
                    resize_round_angle = 1.30
                    if slide_percent_1 >= resize_round_angle:
                        slide_percent_diff = 2.0 - slide_percent_1
                        slide_percent_total = 2.0 - slide_percent_diff
                        slide_percent_diff_percent = slide_percent_diff/slide_percent_total
                        round_offset_diff = self.config.OUTSIDE_ROUND_OFFSET - self.config.ROUND_OFFSET
                        round_offset_diff_added = int(round_offset_diff * slide_percent_diff_percent)
                        round_offset = self.config.ROUND_OFFSET + round_offset_diff_added

                    if not is_apply_large_corner:
                        round_offset = self.config.INSIDE_ROUND_OFFSET

                    format_dict_array, previous_x, previous_y, next_x, next_y = self.make_coner_curve(round_offset,format_dict_array,idx,skip_coordinate_rule)

                    # cache transformed nodes.
                    # we generated nodes
                    # 因為只有作用在2個coordinate. 
                    if self.config.PROCESS_MODE in ["HALFMOON"]:
                        # 加了這行，會讓「口」的最後一個角，無法套到。
                        skip_coordinate.append([previous_x,previous_y])
                        pass

                    check_first_point = True
                    redo_travel=True

                    # current version is not stable!, redo will cuase strange curves.
                    # [BUT], if not use -1, some case will be lost if dot near the first dot.
                    resume_idx = -1
                    #resume_idx = idx
                    break
                    #redo_travel=True
                    #resume_idx = -1
                    #break

        if check_first_point:
            # check close path.
            self.reset_first_point(format_dict_array, spline_dict)


        return redo_travel, resume_idx, inside_stroke_dict,skip_coordinate, skip_coordinate_rule


python/TtfConfig.py
#!/usr/bin/env python3
#encoding=utf-8

class TtfConfig():
    VERSION = "2.125"
    PROCESS_MODE = "GOTHIC"
    #PROCESS_MODE = "HALFMOON"
    #PROCESS_MODE = "D"
    PROCESS_MODE = "B2"
    #PROCESS_MODE = "B4"
    #PROCESS_MODE = "XD"

    STYLE_INDEX = 5
    STYLE_ARRAY = ["Black","Bold","Medium","Regular","DemiLight","Light","Thin"]
    STYLE=STYLE_ARRAY[STYLE_INDEX]

    DEFAULT_COORDINATE_VALUE = -9999

    # for Regular
    STROKE_MAX = 102

    # 有些筆劃很細，有些很粗，設太細似乎又會誤判。
    # 雖然字變重，但因為筆畫複雜時，還是有細的線條。
    STROKE_MIN = 30

    # for X,Y axis equal compare.
    # each 100 px, +- 7 px.
    # PS: 大於等於8，其實滿可怕的。
    EQUAL_ACCURACY_MIN = 3
    EQUAL_ACCURACY_PERCENT = 0.07

    # for Regular
    ROUND_OFFSET = 33

    # don't access the distance within 5
    NEXT_DISTANCE_MIN = 5

    # for Regular
    OUTSIDE_ROUND_OFFSET = 55
    INSIDE_ROUND_OFFSET = 19

    NEED_LOAD_BMP_IMAGE = True

    # some inside block not able to fill 2 curve coner, use small one.
    INSIDE_SMALL_ROUND_OFFSET=15

    # unicode in field
    # 1 to 3
    UNICODE_FIELD = 2

    BMP_PATH = '/Users/chunyuyao/Documents/noto/bmp'


    def apply_weight_setting(self):
        self.STYLE=self.STYLE_ARRAY[self.STYLE_INDEX]

        # for uni53BB, 去, 155
        if self.STYLE=="Black":
            self.STROKE_MAX = 165
        if self.STYLE=="Bold":
            self.STROKE_MAX = 139
        if self.STYLE=="Medium":
            self.STROKE_MAX = 115
        if self.STYLE=="DemiLight":
            self.STROKE_MAX = 96
        if self.STYLE=="Light":
            self.STROKE_MAX = 80
        if self.STYLE=="Thin":
            self.STROKE_MAX = 59

        if self.STYLE=="Black":
            self.STROKE_MIN = 34
        if self.STYLE=="Bold":
            self.STROKE_MIN = 34
        if self.STYLE=="Medium":
            self.STROKE_MIN = 30
        if self.STYLE=="DemiLight":
            self.STROKE_MIN = 27
        if self.STYLE=="Light":
            self.STROKE_MIN = 26
        if self.STYLE=="Thin":
            self.STROKE_MIN = 24

        self.STROKE_ACCURACY_PERCENT = 10
        self.STROKE_WIDTH_MAX = int((self.STROKE_MAX * (100+self.STROKE_ACCURACY_PERCENT))/100)
        self.STROKE_WIDTH_MIN = int((self.STROKE_MIN * (100-self.STROKE_ACCURACY_PERCENT))/100)
        self.STROKE_WIDTH_AVERAGE = int((self.STROKE_MIN+self.STROKE_MAX)/2)
        #print("STROKE_WIDTH_MAX:", STROKE_WIDTH_MAX)
        #print("STROKE_MIN:", STROKE_WIDTH_MIN)


        # 客製化 ROUND_OFFSET
        if self.STYLE=="Black":
            self.ROUND_OFFSET = 48
        if self.STYLE=="Bold":
            self.ROUND_OFFSET = 41
        if self.STYLE=="Medium":
            self.ROUND_OFFSET = 37
        if self.STYLE=="DemiLight":
            self.ROUND_OFFSET = 30
        if self.STYLE=="Light":
            self.ROUND_OFFSET = 27
        if self.STYLE=="Thin":
            self.ROUND_OFFSET = 24

        # 客製化 OUTSIDE_ROUND_OFFSET
        if self.STYLE=="Black":
            self.OUTSIDE_ROUND_OFFSET = 65
        if self.STYLE=="Bold":
            self.OUTSIDE_ROUND_OFFSET = 60
        if self.STYLE=="Medium":
            self.OUTSIDE_ROUND_OFFSET = 60
        if self.STYLE=="DemiLight":
            self.OUTSIDE_ROUND_OFFSET = 50
        if self.STYLE=="Light":
            self.OUTSIDE_ROUND_OFFSET = 40
        if self.STYLE=="Thin":
            self.OUTSIDE_ROUND_OFFSET = 30

        # bigger curve.
        if self.PROCESS_MODE in ["TOOTHPASTE"]:
            if self.STYLE=="Black":
                self.OUTSIDE_ROUND_OFFSET = 75
            if self.STYLE=="Bold":
                self.OUTSIDE_ROUND_OFFSET = 70
            if self.STYLE=="Medium":
                self.OUTSIDE_ROUND_OFFSET = 65
            if self.STYLE=="DemiLight":
                self.OUTSIDE_ROUND_OFFSET = 60
            if self.STYLE=="Light":
                self.OUTSIDE_ROUND_OFFSET = 50
            if self.STYLE=="Thin":
                self.OUTSIDE_ROUND_OFFSET = 35

        # 只需要大彎.
        if self.PROCESS_MODE in ["D","XD","HALFMOON","NUT8","TOOTHPASTE","ALIAS","SPIKE"]:
            self.ROUND_OFFSET=self.OUTSIDE_ROUND_OFFSET

        # 客製化 INSIDE_ROUND_OFFSET
        if self.STYLE=="Black":
            self.INSIDE_ROUND_OFFSET = 29
        if self.STYLE=="Bold":
            self.INSIDE_ROUND_OFFSET = 26
        if self.STYLE=="Medium":
            self.INSIDE_ROUND_OFFSET = 23
        if self.STYLE=="DemiLight":
            self.INSIDE_ROUND_OFFSET = 19
        if self.STYLE=="Light":
            self.INSIDE_ROUND_OFFSET = 19
        if self.STYLE=="Thin":
            self.INSIDE_ROUND_OFFSET = 19

        if self.PROCESS_MODE in ["B2"]:
            self.NEED_LOAD_BMP_IMAGE = False
            self.INSIDE_ROUND_OFFSET = 30

            if self.STYLE=="Black":
                self.INSIDE_ROUND_OFFSET = 45
            if self.STYLE=="Bold":
                self.INSIDE_ROUND_OFFSET = 40
            if self.STYLE=="Medium":
                self.INSIDE_ROUND_OFFSET = 35

        if self.PROCESS_MODE in ["B4"]:
            self.NEED_LOAD_BMP_IMAGE = False
            self.INSIDE_ROUND_OFFSET = 25

            if self.STYLE=="Black":
                self.INSIDE_ROUND_OFFSET = 40
            if self.STYLE=="Bold":
                self.INSIDE_ROUND_OFFSET = 35
            if self.STYLE=="Medium":
                self.INSIDE_ROUND_OFFSET = 30


        # for Regular
        if self.PROCESS_MODE in ["TOOTHPASTE"]:
            self.OUTSIDE_ROUND_OFFSET = 65
            self.INSIDE_ROUND_OFFSET = 25

    def __init__(self, weight_code):
        import datetime

        self.STYLE_INDEX = int(weight_code)
        self.apply_weight_setting()
        print("Transform Mode:", self.PROCESS_MODE)
        print("Transform Style:", self.STYLE)
        print("Transform Version:", self.VERSION)
        print("Transform Time:", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))

    def hello(self):
        print("world!")